Update README (#748)
kubevirt/kubevirt.github.io: README.md

# Contributing to the KubeVirt.io Website

[![Build Status](https://travis-ci.org/kubevirt/kubevirt.github.io.svg?branch=master)](https://travis-ci.org/kubevirt/kubevirt.github.io)

The KubeVirt.io website is a Jekyll driven site hosted by GitHub Pages.

> We gladly welcome contributions or please reach out if you happen to have an idea or find an issue with our contents.  [Here's our guidelines for contents](GUIDELINES.md).

Author: ok

## Getting Started

### Git semantics

We follow a semi standard process to move code from development to production

The basic process is ...

> ```bash

Author: We're good here.  The markdown processors used here is GitHub only

fork -> clone -> branch -> commit -> push -> pull request -> approve -> merge
```

#### fork
This is easy.  Just hit the fork button at the top of the page.  This should be the only time you ever have to do this if you maintain your clone and branches correctly.

#### Clone
Web browse to https://www.github.com/*you*/kubevirt.github.io.  Hit the Code button.  Then pick your clone type.  Personally I do everything via ssh.  Copy the url to your clipboard.  Switch to a terminal and type ...

```bash
git clone *clipboard paste url*
```

#### Remotes

> By default your local git repo will have an `origin` remote pointing to your fork in GitHub.  Since our repos have many contributors you will need to add a remote to periodically synchronize to the project.  We typically call this remote `upstream`.
> 
> To add an additional remote perform the following command...
> 
> ```bash
> git remote add upstream http/ssh url
> ```

Reviewer: ```suggestion
Your local git repo will have a remote alias called `origin` which points to your fork in GitHub.  Our repositories have many contributors and you will need to periodically synchronize your local copy and forked copy with the original projects repository.
The repository of the original project is typically called `upstream`.  Adding a remote alias to the original project can make it easier to keep the forked repository and local copies synchronized.

Add an additional remote alias called upstream using the following command:

```bash
git remote add upstream clone_line_from_the_code_button_earlier
```
```

And then you should see something like...

```bash
$ git remote -v
origin	git@github.com:mazzystr/kubevirt.github.io.git (fetch)
origin	git@github.com:mazzystr/kubevirt.github.io.git (push)
upstream	git@github.com:kubevirt/kubevirt.github.io.git (fetch)
upstream	git@github.com:kubevirt/kubevirt.github.io.git (push)
```

> To sync from upstream I do the following ... but be careful here.  This slams the door hard on existing code. You can lose data.

Reviewer 2: A less familiar way to write:

Note this will overwrite any files in the master branch. Make sure all work is done in local branches, and leave the master branch only for tracking the upstream repository.

Reviewer: I would rewrite this as @cwilkers suggested.

```bash
git checkout master; git fetch upstream; git reset --hard upstream/master; git push origin master -f
```

And since we work from the `source` branch at this time you will need an additional step to sync that branch as well ...

```bash
git checkout source; git fetch upstream; git reset --hard upstream/source; git push origin source -f
```

#### Feature branch
Even though we will accept changes from `source` it is inadvisable and can cause confusion and possibly data loss.  Please use feature branches branched from `source`.  The following command can help ...

```bash
git checkout source; git fetch upstream; git reset --hard upstream/source; git push origin source -f; git branch feat_branch; git checkout feat_branch; git push --set-upstream origin feat_branch
```

> #### rebase

Reviewer: ```suggestion
#### Rebase
```

```bash
git checkout master; git fetch upstream; git reset --hard upstream/master; git push origin master -f
```

*Note* Master branch is purely cosmetic for this repo.  Merges to master **ARE NOT ACCEPTED**.

All work must be branched from `source` branch at this time. Perform the following to sync from upstream ...

```bash
git checkout source; git fetch upstream; git reset --hard upstream/source; git push origin source -f
```

#### Feature branch
Even though changes from a local `source` branch are accepted it is inadvisable, can cause confusion and possibly data loss. Please use feature branches branched from `source` by running the following ...

```bash
git checkout source; git fetch upstream; git reset --hard upstream/source; git push origin source -f; git branch feat_branch; git checkout feat_branch; git push --set-upstream origin feat_branch
```

#### Rebase

Periodically a feature branch will need to be rebased as the local and origin fall behind upstream. Perform the following to rebase ...

```bash
git checkout source; git fetch upstream; git reset --hard upstream/source; git push origin source -f; git checkout feat_branch; git rebase origin/source; git push -f
```

> There is always a strong possibility for merge conflicts. Proceed with caution in resolving. Each conflict must be hand editted. Perform the following to resolve each conflict ...

Reviewer 2: Edited

Reviewer 2: Looks good after one more spell check to catch that "editted"

# Contributing to KubeVirt.io

[![Build Status](https://travis-ci.org/kubevirt/kubevirt.github.io.svg?branch=master)](https://travis-ci.org/kubevirt/kubevirt.github.io)

The [kubevirt.io](https://kubevirt.io) website is a [Jekyll](https://jekyllrb.com/) driven site hosted by GitHub Pages.

Contributions to the KubeVirt website are very much welcomed! Please reach out with ideas for new content or issues with existing content!

## Getting Started

### Git semantics

A semi standard process is followed to move code from development to production.

The basic process is ...

```bash
fork -> clone -> branch -> commit -> push -> pull request -> approve -> merge
```

#### Fork

Create a forked copy of the repository in your own account by pressing the `Fork` button at the top of the repository page. This should be the only time a repository fork needs to be created as long as the fork is properly maintained by performing branch sync and rebase with upstream periodically.

#### Clone

> To clone the forked repository locally web browse to https://www.github.com/*you*/kubevirt.github.io. Click the `Code` button. Select the clone method and copy the url to clipboard. Open a terminal window and  ...

Reviewer 2: ```suggestion
To clone the forked repository locally, browse to your fork at https://www.github.com/*you*/kubevirt.github.io. Click the `Code` button. Select the clone method and copy the url to clipboard. Open a terminal window and  ...
```

```bash
git clone *clipboard paste url*
```

#### Remotes

> By default the local git repo will have a remote alias called `origin` that points to your fork in GitHub. KubeVirt repositories have many contributors  work needs to be done periodically to synchronize local and origin branches with upstream branches. This is typically call this remote `upstream`.

Reviewer 2: ```suggestion
By default the local git repo will have a remote alias called `origin` that points to your fork in GitHub. KubeVirt repositories have many contributors so work needs to be done periodically to synchronize local and origin branches with upstream branches. To enable this work flow on your local clone, add a new remote to the repository. By convention, this remote is named `upstream`.
```

To add an additional remote perform the following command...

```bash
git remote add upstream http/ssh url
```

And then you should see something like...

```bash
$ git remote -v
origin	git@github.com:mazzystr/kubevirt.github.io.git (fetch)
origin	git@github.com:mazzystr/kubevirt.github.io.git (push)
upstream	git@github.com:kubevirt/kubevirt.github.io.git (fetch)
upstream	git@github.com:kubevirt/kubevirt.github.io.git (push)
```

> To sync upstream master branch from upstream perform the following ...

Reviewer 2: ```suggestion
To sync the master branch from the upstream repository, perform the following ...
```

Periodically a feature branch will need to be rebased as the local and origin fall behind upstream. Perform the following to rebase ...

```bash
git checkout source; git fetch upstream; git reset --hard upstream/source; git push origin source -f; git checkout feat_branch; git rebase origin/source; git push -f
```

There is always a strong possibility for merge conflicts. Proceed with caution in resolving. Each conflict must be hand edited. Perform the following to resolve each conflict ...

```bash
# Modify and save the filename
git add filename; git rebase --continue
```

#### Work

[Here is our guidelines for content contribution](GUIDELINES.md).

Each section of the site is broken out into their respective folders.

* `./pages` : website
* `./blogs` : blog posts
* `./docs` : documentation.
* `./videos` : videos

All site images are located under `./assets/images`. Please do not edit these images.

Markdown for blog posts are located under `./posts`. Please follow the existing filename scheme.

Images related to blog entries get placed under `./assets/images/BLOG_POST_TITLE`.
The **BLOG_POST_TITLE** should match the name of the markdown file created under `/_posts`.

> If you are a UI/UX developer help with the structure and layout of the website is definitely needed. Feel free to browse [website issues](https://github.com/kubevirt/kubevirt.github.io/issues?q=is%3Aopen+is%3Aissue+label%3Akind%2Fwebsite) or contribute other ideas.

Reviewer 2: ```suggestion
If you are a UI/UX developer, the structure and layout of the website would greatly benefit from your attention. Feel free to browse [website issues](https://github.com/kubevirt/kubevirt.github.io/issues?q=is%3Aopen+is%3Aissue+label%3Akind%2Fwebsite) or contribute other ideas.
```

## Test work

> The ability to run the same tests locally as in CI is available via the included Makefile. This saves time from having to go through the process just to find out the pull req has a problem that prevents merge.

Reviewer 2: ```suggestion
The Makefile at the base of this repository provides editors the ability to locally run the same tests CI uses to validate changes. This saves time over waiting for online CI resources to spin up just to find out a pull request has a problem that prevents merge.
```

1) Build the test image locally
```bash
$ make build_img
```

**NOTE** If you use `docker` you may need to set `CONTAINER_ENGINE` and `BUILD_ENGINE`:

```bash
$ export CONTAINER_ENGINE=docker
$ export BUILD_ENGINE=docker
```

**NOTE** If you are in an SELinux enabled OS you need to set `SELINUX_ENABLED`:

```bash
$ export SELINUX_ENABLED=True
```

2) Validate page rendering
```bash
make run
```
Open your web browser to http://0.0.0.0:4000

3) Test all hyperlinks
```bash
make test_links
```

4) Test spelling
```bash
make test_spelling
```

> If you discover an error that should not be an error feel free to add the offending word to the dictionary file located at GitHub repo `kubevirt/project-infra/images/yaspeller/.yaspeller.json`

Reviewer 2: ```suggestion
If you discover a flagged spelling error that you believe is not a mistake, feel free to add the offending word to the dictionary file located at GitHub repo `kubevirt/project-infra/images/yaspeller/.yaspeller.json`. Try to keep the dictionary file well ordered, and employ regular expressions to handle plurals and typical conjugations.
```

#### Make sure all tests pass before committing!

#### Submitting your code

1) Commit your code and sign your commits!
```bash
git commit -s -m "The commit message" file1 file 2 file3 ...
```

**Signature verification on commits are required! No exceptions!**

You will see the following in the transaction log
```bash
git log
commit hashbrowns (HEAD -> feat_branch, upstream/source, origin/source, source)
Author: KubeVirt contributer <kubevirt_contributer@kubevirt.io>
Date:   Mon Jan 01 00:00:00 2021 -0700

<your commit message>

Signed-off-by: <your configured git identity>
```

2) Browse to `https://www.github.com/*you*/kubevirt.github.io`

3) Often you will see a `Compare & Pull Request` button ... Click on that

4) Ensure your base branch is `source`, your compare branch is `feat_branch`, and the file diff's are correct.

> 5) Create a nice subject and body for the pull request and click the "Create pull request" button.

Reviewer 2: ```suggestion
5) Create a nice subject and body for the pull request, be sure to tag any related issues, and click the "Create pull request" button.
```

All site images are located under `./assets/images`. Please do not edit these images.

Images that relate to blog entries are located under `./assets/images/BLOG_POST_TITLE`.
The **BLOG_POST_TITLE** should match the name of the markdown file that you added under `/_posts`.

If you are a UI/UX developer we definitely need help with the structure of the website.  Feel free to browse [website issues](https://github.com/kubevirt/kubevirt.github.io/issues?q=is%3Aopen+is%3Aissue+label%3Akind%2Fwebsite) or contribute other ideas.

## Test your work

> We have the ability to run the same  tests locally as in CI.  This saves time from having to go through the  process just to find out the pull_req has a problem that prevents merge.

Reviewer: ```suggestion
We have the ability to run the same tests locally that are run in CI.  This saves time from having to go through the  process just to find out the pull_req has a problem that prevents merge.
```
